### ISSUE TITLE
[Phase 3] Model Implementation - CNN Architectures

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/gravitational-wave-detection#4]
## Objective

Implement deep learning architectures for gravitational wave signal detection.

**Part of:** #1

## Dependencies

- #2 (Data Infrastructure) complete
- #3 (Feature Engineering) complete

### Base Model Interface
- [ ] Create `src/models/base.py` with abstract base class
- [ ] Define fit, predict, evaluate interface

### 1D CNN
- [ ] Create `src/models/cnn1d.py` with `GWDetector1D`
- [ ] Multi-branch for 3 detectors
- [ ] Binary classification output

### 2D CNN (Optional)
- [ ] Create `src/models/cnn2d.py` for spectrogram input
- [ ] ResNet-style architecture

### Training Loop
- [ ] Create `src/models/trainer.py`
- [ ] Early stopping, learning rate scheduling
- [ ] Mixed precision training (optional)

### 1D CNN Architecture

```python
# src/models/cnn1d.py
import torch
import torch.nn as nn

class GWDetector1D(nn.Module):
    def __init__(self, n_detectors=3, sample_length=4096):
        super().__init__()

        # Per-detector feature extraction
        self.detector_convs = nn.ModuleList([
            nn.Sequential(
                nn.Conv1d(1, 32, kernel_size=7, padding=3),
                nn.BatchNorm1d(32),
                nn.ReLU(),
                nn.MaxPool1d(2),
                nn.Conv1d(32, 64, kernel_size=5, padding=2),
                nn.BatchNorm1d(64),
                nn.ReLU(),
                nn.MaxPool1d(2),
                nn.Conv1d(64, 128, kernel_size=3, padding=1),
                nn.BatchNorm1d(128),
                nn.ReLU(),
                nn.AdaptiveAvgPool1d(16)
            )
            for _ in range(n_detectors)
        ])

        # Combined classifier
        self.classifier = nn.Sequential(
            nn.Linear(128 * 16 * n_detectors, 256),
            nn.ReLU(),
            nn.Dropout(0.5),
            nn.Linear(256, 64),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(64, 1),
            nn.Sigmoid()
        )

    def forward(self, x):
        # x: (batch, 3, 4096)
        features = []
        for i, conv in enumerate(self.detector_convs):
            det_x = x[:, i:i+1, :]  # (batch, 1, 4096)
            feat = conv(det_x)  # (batch, 128, 16)
            features.append(feat.flatten(1))

        combined = torch.cat(features, dim=1)
        return self.classifier(combined)
```

### Training Loop

```python
# src/models/trainer.py
import torch
from torch.optim import AdamW
from torch.nn import BCELoss

class ModelTrainer:
    def __init__(self, model, lr=1e-3, device='cuda'):
        self.model = model.to(device)
        self.device = device
        self.optimizer = AdamW(model.parameters(), lr=lr)
        self.criterion = BCELoss()

    def train_epoch(self, dataloader):
        self.model.train()
        total_loss = 0
        for X, y in dataloader:
            X = X.float().to(self.device)
            y = y.float().to(self.device).unsqueeze(1)

            self.optimizer.zero_grad()
            pred = self.model(X)
            loss = self.criterion(pred, y)
            loss.backward()
            self.optimizer.step()

            total_loss += loss.item()
        return total_loss / len(dataloader)
```

## Definition of Done

- [ ] 1D CNN achieves AUC > 0.85 on G2Net validation
- [ ] Training completes in reasonable time (<1 hour for 10K samples)
- [ ] Model saves and loads correctly
- [ ] Loss decreases during training

## Technical Notes

- Use BCEWithLogitsLoss for numerical stability (remove sigmoid from model)
- Mixed precision (AMP) can speed up training significantly
- Consider attention mechanisms for detector fusion
